### DISCUSSION TITLE
Please test version 8.2.2.2 Beta (Firefox only)

### DISCUSSION THREAD

**Discussion by paulrouget** [aclap-dev/video-downloadhelper#248] [Announcements]
A subtle improvement for some videos where we select a better CDN for faster speed (impact vimeo mostly): https://downloadhelper.net/data/vdh-mozilla-beta-latest.xpi

[the remark this answers is not in the document]
**Reply by Mattmjs** [reply]
If you remember on this one https://github.com/aclap-dev/video-downloadhelper/discussions/236#discussioncomment-8639331 it would suffer an error with HLS download. Now it works.

On embedded videos there isn't immediate detection until 1 or 2 reloads.

**Comment by spookycat567** [comment]
Hi, Is there one for chrome?

**Reply by paulrouget** [reply]
No not yet.

[the remark this answers is not in the document]
**Reply by paulrouget** [reply]
@Mattmjs I think I improved the situation with 8.2.2.3. Can you confirm?

~Edit: It seems like it didn't fully resolved the issues. Investigating more…~
Edit: seems fine.

[the remark this answers is not in the document]
**Reply by Mattmjs** [reply]
I tested and it works on embedded videos with detection first time when on an actual webpage.

There are these other videos that are not on any webpage and require a reload before there is detection. 
Example from the google group : https://player.vimeo.com/video/898312290

[the remark this answers is not in the document]
**Reply by GrampaWildWilly** [reply]
I'm on 8.2.2.3a1 now.  I went to the page Matt mentioned & tried opening the gear icon & selecting a resolution.  I tried playing the video for a couple of seconds.  The VDH menu remained empty.  Only a page reload made the VDH menu populate, & I didn't have to click play to make that happen.

[the remark this answers is not in the document]
**Reply by GrampaWildWilly** [reply]
When you say "better CDN," how do you determine that?  My observation has been that with Vimeo content, the master manifests have been in domains akamaized.net & skyfire.vimeocdn.  I have done some experiments (not recently) in which I downloaded the content in question using the master manifests in both domains.  The performance of the downloads was comparable, & the resulting downloaded videos were identical.  So what tests do you apply to choose between the manifests in the 2 domains?  What makes one CDN preferable over the other?

As for HLS vs DASH, I generally have not found DASH master manifests with Vimeo content, only HLS master manifests.  But I must admit I haven't tried Matt's trick of using the HLS master manifest URL & just swapping in mpd for m3u8.  I was always just happy I had a master manifest of any sort & used it.

[the remark this answers is not in the document]
**Reply by paulrouget** [reply]
Took me some time to understand that one. I think this is actually detected, but as a an orphan. Looking at it…

[the remark this answers is not in the document]
**Reply by paulrouget** [reply]
> What makes one CDN preferable over the other?

Here, there's a real difference between the 2 offered CDN, and the best is the one in the config json.

[the remark this answers is not in the document]
**Reply by paulrouget** [reply]
@Mattmjs @GrampaWildWilly I've pushed version 8.2.2.4 which should fix the problem with the last link.

[the remark this answers is not in the document]
**Reply by Mattmjs** [reply]
It works first time, no reload needed. So next stable on the way ? or other things need sorting out first.

[the remark this answers is not in the document]
**Reply by GrampaWildWilly** [reply]
For Matt's URL, there is no config json.  In fact, there is no json at all in the Network Monitor.  The information is in a file of type HTML in the domain player.vimeo.com, the domain where we often find our manifest information.  I did, as we often do, a Save Response.  When I paste that into a Notepad++ window, I get a file that pretty much does look like a json.  But its name isn't config.  It's https://player.vimeo.com/video/898312290.  Inside there, I find 7 m3u8 URLs.  But it turns out there's really only 2 URLs.  The first 4 URLs are duplicates of a manifest in domain akamaized.net.  The last 3 URLs are duplicates of a manifest in domain skyfire.vimeocdn.  This pattern of 7 is something I've seen a number of times before.

So after adjusting % to %% and \u0026 to &, I ran ffprobe on the 2 manifests:

[ffprobe.txt](https://github.com/aclap-dev/video-downloadhelper/files/14493945/ffprobe.txt)

There's 2 ffprobe reports in there.  They look pretty much the same, including the captions.

In this case, the ampersands are escaped as `\u0026` which is not what I usually see.  Usually, I see `\\u0026` with back slash being doubled.  So I don't even know how they decide to do things.

So I ran 4 ffmpeg invocations, 2 for the mp4 videos & 2 for the vtt subtitles.

[Akamaized Video mp4 Log.txt](https://github.com/aclap-dev/video-downloadhelper/files/14493967/Akamaized.Video.mp4.Log.txt)
[Akamaized Video vtt Log.txt](https://github.com/aclap-dev/video-downloadhelper/files/14494157/Akamaized.Video.vtt.Log.txt)
[Skyfire Video mp4 Log.txt](https://github.com/aclap-dev/video-downloadhelper/files/14493977/Skyfire.Video.mp4.Log.txt)
[Skyfire Video vtt Log.txt](https://github.com/aclap-dev/video-downloadhelper/files/14493983/Skyfire.Video.vtt.Log.txt)

Here are the results.  First, the 2 captions:

![#01](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/73a14cca-0a0d-4454-a7c0-665ebf3adf6c)

Now the 2 mp4 videos:

![#02](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/7ed21caa-caae-495d-bf0b-dc306a1c41c4)
![#03](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/dfbaa427-ccb4-4b85-b1f3-5afd72599313)

The akamaized download is somewhat faster, but not by a lot.  In any case, as you can see, the 2 CDNs supplied identical results.

And here is one of them playing in VLC:

![#05](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/e8e2b91d-9cd9-442d-9ff7-4c0bb247ad30)

I didn't sit & watch them, just sampled them at intervals.  They both had audio & video all the way to the end.  They also both had the subtitles.  The ffprobe reports label these captions as machine generated.  I didn't watch this carefully enough to assess how accurate the subtitles are.  I've seen some pretty lousy automatically generated captions & I've seen some decent ones.  Somebody who is actually interested in sitting in front of this for an hour & a half will have to tell us how good these subtitles might be.

But I am still mystified how you decide which one to choose.  There is no json at all with this video.  The required information is in an object identified as HTML.  Its contents aren't even HTML but the manifest URLs are in there.

![#04](https://github.com/aclap-dev/video-downloadhelper/assets/75588964/05becca6-3179-4508-945f-a522169ec7cc)

Based on our experience, there are several patterns with Vimeo content.  We've always just improvised our solutions.  Every case seems to have presented its own template.  There are a few templates.  This one doesn't even supply a json.  How will you choose which CDN to rely on?  If one of them seems consistently better in Taiwan, that doesn't necessarily mean the same one will be better in Toulouse.  How are you making this decision?

[the remark this answers is not in the document]
**Reply by paulrouget** [reply]
> But I am still mystified how you decide which one to choose. There is no json at all with this video

The json is not always there, but accessible in the html. From there there's are info to know which CDN should be the default.

[the remark this answers is not in the document]
**Reply by GrampaWildWilly** [reply]
Oh.  Tricky.  So there is always a json, just not always visible in the Network Monitor.  Something to keep you busy, right?
